File: 0707-design-linked-list/0707-design-linked-list-DoublyLinkedListUsingSentinelNodes.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
    struct Node* prev;  // Pointer to the previous node
    int val;            // Value stored in the node
    struct Node* next;  // Pointer to the next node
} Node;

typedef struct {
    Node* head;    // Dummy head node
    Node* tail;    // Dummy tail node
    int length;    // Length of the linked list (excluding dummy nodes)
} MyLinkedList;

// Helper function to create a new node
Node* createNode(int val) {
    Node* node = (Node*)malloc(sizeof(Node));
    if (node == NULL) {
        fprintf(stderr, "Node Memory Allocation Failed");
        return NULL;
    }
    node->val = val;
    node->prev = NULL;
    node->next = NULL;
    return node;
}

// Function to create a new linked list with dummy nodes
MyLinkedList* myLinkedListCreate() {
    MyLinkedList* list = (MyLinkedList*)malloc(sizeof(MyLinkedList));
    if (list == NULL) {
        fprintf(stderr, "Linked List Memory Allocation Failed");
        return NULL;
    }
    // Initialize dummy nodes
    list->head = createNode(-1);  // Dummy head node with arbitrary value
    list->tail = createNode(-1);  // Dummy tail node with arbitrary value

    if (list->head == NULL || list->tail == NULL) {
        free(list);
        return NULL;
    }

    // Link the dummy nodes to each other
    list->head->next = list->tail;
    list->tail->prev = list->head;

    list->length = 0;  // Initialize length to 0
    return list;
}

// Function to get the value at the given index (0-based)
int myLinkedListGet(MyLinkedList* obj, int index) {
    if (index < 0 || index >= obj->length) {
        return -1;  // Return -1 for an invalid index
    }
    
    Node* currentNode = obj->head->next;  // Start from the first real node (after the dummy head)
    for (int i = 0; i < index; i++) {
        currentNode = currentNode->next;
    }
    return currentNode->val;
}

// Function to add a node with value `val` at the head (index 0)
void myLinkedListAddAtHead(MyLinkedList* obj, int val) {
    Node* newNode = createNode(val);
    if (newNode == NULL) return;

    Node* firstRealNode = obj->head->next;  // The first real node after the dummy head

    // Insert newNode between the head dummy and the first real node
    newNode->next = firstRealNode;
    newNode->prev = obj->head;
    
    obj->head->next = newNode;
    firstRealNode->prev = newNode;

    obj->length++;
}

// Function to add a node with value `val` at the tail
void myLinkedListAddAtTail(MyLinkedList* obj, int val) {
    Node* newNode = createNode(val);
    if (newNode == NULL) return;

    Node* lastRealNode = obj->tail->prev;  // The last real node before the dummy tail

    // Insert newNode between the last real node and the tail dummy
    newNode->next = obj->tail;
    newNode->prev = lastRealNode;
    
    lastRealNode->next = newNode;
    obj->tail->prev = newNode;

    obj->length++;
}

// Function to add a node with value `val` at the given index
void myLinkedListAddAtIndex(MyLinkedList* obj, int index, int val) {
    if (index < 0 || index > obj->length) {
        return;  // Invalid index
    }

    if (index == 0) {
        myLinkedListAddAtHead(obj, val);  // Insert at head
    } else if (index == obj->length) {
        myLinkedListAddAtTail(obj, val);  // Insert at tail
    } else {
        Node* newNode = createNode(val);
        if (newNode == NULL) return;

        Node* currentNode = obj->head->next;  // Start from the first real node
        for (int i = 0; i < index - 1; i++) {
            currentNode = currentNode->next;
        }

        // Insert newNode between currentNode and currentNode->next
        newNode->next = currentNode->next;
        newNode->prev = currentNode;
        
        currentNode->next->prev = newNode;
        currentNode->next = newNode;

        obj->length++;
    }
}

// Function to delete the node at the given index
void myLinkedListDeleteAtIndex(MyLinkedList* obj, int index) {
    if (index < 0 || index >= obj->length) {
        return;  // Invalid index
    }

    Node* currentNode = obj->head->next;  // Start from the first real node
    for (int i = 0; i < index; i++) {
        currentNode = currentNode->next;
    }

    // Update the previous and next pointers to remove the currentNode
    currentNode->prev->next = currentNode->next;
    currentNode->next->prev = currentNode->prev;

    free(currentNode);  // Free the memory of the deleted node
    obj->length--;
}

// Function to free the entire linked list, including dummy nodes
void myLinkedListFree(MyLinkedList* obj) {
    Node* currentNode = obj->head;
    while (currentNode != NULL) {
        Node* nextNode = currentNode->next;
        free(currentNode);  // Free each node
        currentNode = nextNode;
    }
    free(obj);  // Free the linked list structure
}

/**
 * Your MyLinkedList struct will be instantiated and called as such:
 * MyLinkedList* obj = myLinkedListCreate();
 * int param_1 = myLinkedListGet(obj, index);
 * myLinkedListAddAtHead(obj, val);
 * myLinkedListAddAtTail(obj, val);
 * myLinkedListAddAtIndex(obj, index, val);
 * myLinkedListDeleteAtIndex(obj, index);
 * myLinkedListFree(obj);
 */

File: Difficulty/B. Medium/0707-design-linked-list/0707-design-linked-list-DoublyLinkedListUsingSentinelNodes.py
class Node:
    def __init__(self, val=None):
        self.val = val  # The value stored in the node
        self.prev = None  # Pointer to the previous node
        self.next = None  # Pointer to the next node

class MyLinkedList:
    def __init__(self):
        # Create dummy head and tail nodes
        self.head = Node()  # Dummy head node
        self.tail = Node()  # Dummy tail node

        # Initially, link head and tail to each other
        self.head.next = self.tail
        self.tail.prev = self.head

        self.length = 0  # Number of elements in the list (excluding dummy nodes)

    # Get the value of the node at the given index
    def get(self, index: int) -> int:
        if index < 0 or index >= self.length:
            return -1  # Invalid index, return -1

        current = self.head.next  # Start from the first real node
        for _ in range(index):
            current = current.next
        return current.val

    # Add a node with value `val` at the head (index 0)
    def addAtHead(self, val: int) -> None:
        new_node = Node(val)

        # Insert new_node right after the dummy head
        first_real_node = self.head.next
        new_node.next = first_real_node
        new_node.prev = self.head

        self.head.next = new_node
        first_real_node.prev = new_node

        self.length += 1

    # Add a node with value `val` at the tail
    def addAtTail(self, val: int) -> None:
        new_node = Node(val)

        # Insert new_node right before the dummy tail
        last_real_node = self.tail.prev
        new_node.next = self.tail
        new_node.prev = last_real_node

        last_real_node.next = new_node
        self.tail.prev = new_node

        self.length += 1

    # Add a node with value `val` at the given index
    def addAtIndex(self, index: int, val: int) -> None:
        if index < 0 or index > self.length:
            return  # Invalid index, do nothing

        if index == 0:
            self.addAtHead(val)  # Special case: Insert at the head
        elif index == self.length:
            self.addAtTail(val)  # Special case: Insert at the tail
        else:
            new_node = Node(val)

            # Traverse to the node before the insertion point
            current = self.head.next
            for _ in range(index - 1):
                current = current.next

            # Insert new_node between current and current.next
            next_node = current.next
            new_node.next = next_node
            new_node.prev = current

            current.next = new_node
            next_node.prev = new_node

            self.length += 1

    # Delete the node at the given index
    def deleteAtIndex(self, index: int) -> None:
        if index < 0 or index >= self.length:
            return  # Invalid index, do nothing

        # Traverse to the node to delete
        current = self.head.next
        for _ in range(index):
            current = current.next

        # Update pointers to remove current node
        prev_node = current.prev
        next_node = current.next

        prev_node.next = next_node
        next_node.prev = prev_node

        self.length -= 1

    # Display list (for debugging purposes)
    def display(self):
        current = self.head.next
        elements = []
        while current != self.tail:
            elements.append(current.val)
            current = current.next
        print("List:", elements)

# Your MyLinkedList object will be instantiated and called as such:
# obj = MyLinkedList()
# param_1 = obj.get(index)
# obj.addAtHead(val)
# obj.addAtTail(val)
# obj.addAtIndex(index,val)
# obj.deleteAtIndex(index)